<a href="https://colab.research.google.com/github/francomanca93/analisis-de-datos/blob/funcionalidades-basicas/2-Funcionalidades-basicas-de-pandas/7_Tipos_de_datos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tipos de Variables que componen un data frame

Cual es la rutina de preprocesamiento de datos con la que debemos comenzar a analizar un dataframe. Vamos a estudiar la composicion de las diferentes formatos de variables que podemos encontrar en un dataframe. 


Vamos a utiilzar [Google DataSearch](https://datasetsearch.research.google.com/) para buscar datasets.


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [2]:
%cd '/content/drive/My Drive/Colab Notebooks/analisis-de-datos-con-pandas/db'
!ls

/content/drive/My Drive/Colab Notebooks/analisis-de-datos-con-pandas/db
Meteorite_Landings.csv	test.hdf      test.pkl
test.csv		test.json     test-separacion-diferente.csv
test.gsheet		test.parquet  test.xlsx


In [3]:
import pandas as pd

In [4]:
df_meteorites = pd.read_csv('Meteorite_Landings.csv')

In [5]:
df_meteorites.head() # Por defecto las primeras 5 filas. Podemos ver mas

name   id nametype  ...    reclat    reclong           GeoLocation
0    Aachen    1    Valid  ...  50.77500    6.08333     (50.775, 6.08333)
1    Aarhus    2    Valid  ...  56.18333   10.23333  (56.18333, 10.23333)
2      Abee    6    Valid  ...  54.21667 -113.00000    (54.21667, -113.0)
3  Acapulco   10    Valid  ...  16.88333  -99.90000     (16.88333, -99.9)
4   Achiras  370    Valid  ... -33.16667  -64.95000   (-33.16667, -64.95)

[5 rows x 10 columns]

In [6]:
df_meteorites.tail(10) # Para ver la cola, ultimos registros.

name     id nametype  ...    reclat    reclong             GeoLocation
45706     Zerkaly  31354    Valid  ...  52.13333   81.96667    (52.13333, 81.96667)
45707    Zhaoping  54609    Valid  ...  24.23333  111.18333   (24.23333, 111.18333)
45708    Zhigansk  30405    Valid  ...  68.00000  128.30000           (68.0, 128.3)
45709  Zhongxiang  30406    Valid  ...  31.20000  112.50000           (31.2, 112.5)
45710  Zillah 001  31355    Valid  ...  29.03700   17.01850       (29.037, 17.0185)
45711  Zillah 002  31356    Valid  ...  29.03700   17.01850       (29.037, 17.0185)
45712      Zinder  30409    Valid  ...  13.78333    8.96667     (13.78333, 8.96667)
45713        Zlin  30410    Valid  ...  49.25000   17.66667       (49.25, 17.66667)
45714   Zubkovsky  31357    Valid  ...  49.78917   41.50460     (49.78917, 41.5046)
45715  Zulu Queen  30414    Valid  ...  33.98333 -115.68333  (33.98333, -115.68333)

[10 rows x 10 columns]

In [7]:
df_meteorites.sample(10) # Para ver una muestra aleatoria

name     id  ...     reclong              GeoLocation
42690                  Yamato 8320  29425  ...   35.666670        (-71.5, 35.66667)
42788                 Yamato 86019  29525  ...   35.666670        (-71.5, 35.66667)
36116         Roberts Massif 04120  44588  ...         NaN                      NaN
9555        Elephant Moraine 83317   7956  ...  157.230000       (-76.2975, 157.23)
31096                        Ozren  18067  ...   18.418060      (44.6125, 18.41806)
14283       Grove Mountains 022217  47922  ...   75.320833  (-72.778333, 75.320833)
21221            Lewis Cliff 88495  14239  ...  161.387470   (-84.25064, 161.38747)
26986        Northwest Africa 1851  17508  ...         NaN                      NaN
32369  Queen Alexandra Range 93115  19204  ...  162.534600    (-84.63091, 162.5346)
30434        Northwest Africa 7348  55716  ...    0.000000               (0.0, 0.0)

[10 rows x 10 columns]

In [8]:
df_meteorites.shape # Tamaño de la base de datos, Muestras las filas y las columnas

(45716, 10)

In [9]:
df_meteorites.size # filas x columnas. Registros totales

457160

In [10]:
df_meteorites.describe()

id      mass (g)        reclat       reclong
count  45716.000000  4.558500e+04  38401.000000  38401.000000
mean   26889.735104  1.327808e+04    -39.122580     61.074319
std    16860.683030  5.749889e+05     46.378511     80.647298
min        1.000000  0.000000e+00    -87.366670   -165.433330
25%    12688.750000  7.200000e+00    -76.714240      0.000000
50%    24261.500000  3.260000e+01    -71.500000     35.666670
75%    40656.750000  2.026000e+02      0.000000    157.166670
max    57458.000000  6.000000e+07     81.166670    354.473330

Podemos quitar la notacion cientifica con la siguiente linea.

In [11]:
pd.options.display.float_format = '{:,.1f}'.format

In [12]:
df_meteorites.describe()

id     mass (g)   reclat  reclong
count 45,716.0     45,585.0 38,401.0 38,401.0
mean  26,889.7     13,278.1    -39.1     61.1
std   16,860.7    574,988.9     46.4     80.6
min        1.0          0.0    -87.4   -165.4
25%   12,688.8          7.2    -76.7      0.0
50%   24,261.5         32.6    -71.5     35.7
75%   40,656.8        202.6      0.0    157.2
max   57,458.0 60,000,000.0     81.2    354.5

Interpretacion en los cuartiles

Si ordenamos los datos de menor a mayor:
* Al cortar la base en el primer 25% obtendremos que la masa de los meteoritos en ese corte es de 7.2 g
* Al cortarlo en el 50% quiere decir que el valor del peso que se encuentra exactamente a la mitad, es decir, en ese corte es de 32.6 g. 

Otra explicación que podríamos tener aqui es que el 50% de meteoritos no tienen un peso mayor a 32.6 g y asi sucesivamente con el último cuartil.

In [13]:
df_meteorites.describe(include='all')

name       id nametype  ...   reclat  reclong GeoLocation
count          45716 45,716.0    45716  ... 38,401.0 38,401.0       38401
unique         45716      nan        2  ...      nan      nan       17100
top     Yamato 74146      nan    Valid  ...      nan      nan  (0.0, 0.0)
freq               1      nan    45641  ...      nan      nan        6214
mean             NaN 26,889.7      NaN  ...    -39.1     61.1         NaN
std              NaN 16,860.7      NaN  ...     46.4     80.6         NaN
min              NaN      1.0      NaN  ...    -87.4   -165.4         NaN
25%              NaN 12,688.8      NaN  ...    -76.7      0.0         NaN
50%              NaN 24,261.5      NaN  ...    -71.5     35.7         NaN
75%              NaN 40,656.8      NaN  ...      0.0    157.2         NaN
max              NaN 57,458.0      NaN  ...     81.2    354.5         NaN

[11 rows x 10 columns]

In [14]:
df_meteorites.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45716 entries, 0 to 45715
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         45716 non-null  object 
 1   id           45716 non-null  int64  
 2   nametype     45716 non-null  object 
 3   recclass     45716 non-null  object 
 4   mass (g)     45585 non-null  float64
 5   fall         45716 non-null  object 
 6   year         45425 non-null  object 
 7   reclat       38401 non-null  float64
 8   reclong      38401 non-null  float64
 9   GeoLocation  38401 non-null  object 
dtypes: float64(3), int64(1), object(6)
memory usage: 3.5+ MB


In [15]:
df_meteorites.dtypes

name            object
id               int64
nametype        object
recclass        object
mass (g)       float64
fall            object
year            object
reclat         float64
reclong        float64
GeoLocation     object
dtype: object

In [16]:
df_meteorites.convert_dtypes().dtypes

name            string
id               Int64
nametype        string
recclass        string
mass (g)       float64
fall            string
year            string
reclat         float64
reclong        float64
GeoLocation     string
dtype: object